### DISCUSSION TITLE
plugin.xml  creating userClasses that are managed only by Main Admin or designated userClass?

### DISCUSSION THREAD

**Discussion by Vodhin** [e107inc/e107#5009] [Q&A Plugins]
I am creating a Real Estate plugin that can handle multiple property listings. What controls each listing is the Estate Company, which have their own Agency Locations to which individual Agents are assigned (as a home base). I am assuming that someone who adopts this plugin would likely be manually adding Users by creating a login profile for each of their employees with Admin and Estate Plugin permissions, and the Main Admin doing something similar to onboard Companies.

The plugin is capable of handling multiple companies, so it requires some user credentials to prevent cross-company access. I would like my plugin to create three special user classes when installed, which are used by my plugin to determine what a user can do: 

- Users with the **estate_agent** class can add/edit/delete their own listings. 
- Users with the **estate_manager** class can do all the above for any listings within the company as well as add/edit/delete Agency Locations and assign Users to the estate_agent class. 
- Users with the **estate_admin** class can do all of the above as well as add/edit/delete companies.

Each of these classes need to be regular Admins with permissions to access the Estate Plugin, and I would like the estate_admin to be able to assign the estate_manager and estate_agent class to any User, and the estate_manager to be able to assign the estate_agent class to any User (understanding that each User would need to be made an Admin with Estate Plugin access).

I've added the following to the `<userClasses>` tag in my plugin.xml

`<class name="estate_admin" description="Estate Plugin Administrator" />`
`<class name="estate_manager" description="Estate Agency Manager" />`
`<class name="estate_agent" description="Estate Agent" />`

which works fine for creating the user classes but by default they are visible to and selectable by any member in their own user settings page. I've looked though the plugin_class.php file and it seems that no other attributes can be added through the plugin.xml file.

It would be great to be able to do something like: 
`<class name="estate_agent" manager="250" visibility="250" description="Estate Admin" />`

or even 
`<class name="estate_admin" manager="250" visibility="250" description="Estate Agent" />`
`<class name="estate_manager"  manager="250,estate_admin" visibility="250,estate_admin" description="Estate Agency Manager" />`
`<class name="estate_agent" manager="250,estate_admin,estate_manager" visibility="250,estate_admin,estate_manager" description="Estate Agent" />`

**Comment by Jimmi08** [comment]
HI, I don't have an answer,  I didn't know that you can add classes by plugin.xml either. But I needed this too and I solved it by using plugin_setup.php file, install_post() method, and SQL query to add or update this stuff.

**Reply by Vodhin** [reply]
I did look into plugin_setup.php but wondered what would happen if the userclass_id was defined as 0 (examples seem to have a fixed value) or, if given a numerical value, what would happen if the ID already existed?

**Reply by Jimmi08** [reply]
Not my area, I haven't looked into this but generally, I avoid using plugin.xml if I can.  There are no checks under your control, it just overrides everything. For simple stuff, it works nicely, but I am building  efiction plugin and ...  wait when you get to user extended fields.  I could select between custom table (like for next fields for agents) or UEF. I decided for inbuild stuff and... not sure now if it was so good solution.  Some things work awesome but some are overcomplicated.  
And I do must of the stuff in e_module addon. 

By the way, if you want the answer from the core developer, maybe the issue is the better idea.

**Reply by Vodhin** [reply]
Thanks for the reply. I'll post a stripped down version of this question in the Issues section. 

I hear you on User Extended fields. I tried them once way back in 2006 but found it easier to just create my own table with the needed fields and link those records by USERID . This too can become over complicated, especially since I tend to be heavy-handed when I develop my code. I always think I'll trim the fat sometime later, and I usually do but only when I figure out what I should be doing, which doesn't happen as much as it should (Heck, I just started to figure out how PHP Classes work after 18 years of tinkering with e107!)
